#### CSCE 670 :: Information Storage & Retrieval :: Texas A&M University :: Spring 2023


# Homework 2: Implicit Recommender Systems

### 100 points [9% of your final grade]

### Due: March 9 (Thursday) by 11:59pm

*Goals of this homework:* The objective of this homework is to get you familiar with the pipelines of implicit recommendation, turn the theories of collaborative filtering, matrix factorization, and bayesian personalized ranking (BPR) into practice, and compare the results with the naive non-personalized recommendation to see how personalized collaborative filtering algorithms, matrix factorization model, and the BPR model provide better recommendations for the ranking task (with implicit feedback).

*Submission instructions (eCampus):* To submit your homework, rename this notebook as `UIN_hw1.ipynb`. For example, my homework submission would be something like `555001234_hw1.ipynb`. Submit this notebook via eCampus (look for the homework 2 assignment there). Your notebook should be completely self-contained, with the results visible in the notebook. We should not have to run any code from the command line, nor should we have to run your code within the notebook (though we reserve the right to do so). So please run all the cells for us, and then submit.

*Late submission policy:* For this homework, you may use as many late days as you like (up to the 5 total allotted to you).

*Collaboration policy:* You are expected to complete each homework independently. Your solution should be written by you without the direct aid or help of anyone else. However, we believe that collaboration and team work are important for facilitating learning, so we encourage you to discuss problems and general problem approaches (but not actual solutions) with your classmates. You may post on Piazza, search StackOverflow, etc. But if you do get help in this way, you must inform us by **filling out the Collaboration Declarations at the bottom of this notebook**. 

*Example: I found helpful code on stackoverflow at https://stackoverflow.com/questions/11764539/writing-fizzbuzz that helped me solve Problem 2.*

The basic rule is that no student should explicitly share a solution with another student (and thereby circumvent the basic learning process), but it is okay to share general approaches, directions, and so on. If you feel like you have an issue that needs clarification, feel free to contact either me or the TA.

# Part 1. Recommendations with Implicit Feedback (20 points)

For this part, we're going to build a simple **non-personalized** implicit recommendation algorithm. Since feedback like user clicks, purchases, and views is much more widespread than explicit ratings, implicit recommenders offer great opportunities for far-reaching impact. 

Concretely, the task of implicit recommendation is to recommend items to users based on implicit signals from users, i.e., we only know what items a user is interested in, but have no idea what items the user dislikes. So for this case, the dataset we could use for this implicit recommendation experiment only contains binary data with 1 representing that the user likes the item, and with 0 representing that we don't know the user's preference towards the item. Because of this, we cannot use the same evaluation method as explicit recommendation. Instead, we need to evaluate the implicit recommendation quality by a ranking task.

For this part, we will:
* load and process the MovieLens 1M dataset,
* transfer the explicit dataset to implicit one,
* build a non-personalized implicit recommender, 
* and evaluate your recommender.

To start out, we need to prepare the data. We will use the MovieLens 1M data from https://grouplens.org/datasets/movielens/1m/ in this homework. Lucky for you, we are providing the file containing the ratings -- ratings.dat  -- so all you need to do is load the ratings.dat file in the notebook as a DataFrame variable using the Pandas library. The code to do this has been provided in the next cell, but you need to run it. The resulting data variables are: train_mat is the numpy array variable for training data of size (#users, #items) with non-zero entries representing user-item ratings, and zero entries representing unknown user-item ratings; and test_mat is the numpy array variable for testing data of size (#users, #items).

In [3]:
import pandas as pd
import numpy as np
from scipy.sparse import coo_matrix

np.random.seed(0)

data_df = pd.read_csv('./ratings.dat', sep='::', names=["UserID", "MovieID", "Rating", "Timestamp"], engine='python')

# First, generate dictionaries for mapping old id to new id for users and movies
unique_MovieID = data_df['MovieID'].unique()
unique_UserID = data_df['UserID'].unique()
j = 0
user_old2new_id_dict = dict()
for u in unique_UserID:
    user_old2new_id_dict[u] = j
    j += 1
j = 0
movie_old2new_id_dict = dict()
for i in unique_MovieID:
    movie_old2new_id_dict[i] = j
    j += 1
    
# Then, use the generated dictionaries to reindex UserID and MovieID in the data_df
user_list = data_df['UserID'].values
movie_list = data_df['MovieID'].values
for j in range(len(data_df)):
    user_list[j] = user_old2new_id_dict[user_list[j]]
    movie_list[j] = movie_old2new_id_dict[movie_list[j]]
data_df['UserID'] = user_list
data_df['movieID'] = movie_list

# generate train_df with 70% samples and test_df with 30% samples, and there should have no overlap between them.
train_index = np.random.random(len(data_df)) <= 0.7
train_df = data_df[train_index]
test_df = data_df[~train_index]

# generate train_mat and test_mat
num_user = len(data_df['UserID'].unique())
num_movie = len(data_df['MovieID'].unique())

train_mat = coo_matrix((train_df['Rating'].values, (train_df['UserID'].values, train_df['MovieID'].values)), shape=(num_user, num_movie)).astype(float).toarray()
test_mat = coo_matrix((test_df['Rating'].values, (test_df['UserID'].values, test_df['MovieID'].values)), shape=(num_user, num_movie)).astype(float).toarray()

It is very easy to transfer the explicit datasets you already generated to implicit ones: here, you will consider the watching behavior as the implicit signal showing that the user is interested in a movie. Thus, you can use the same train_df and test_df for implicit recommendation experiment with 'Rating' column ignored. And for train_mat and test_mat, you need to make all ratings to be value 1 and keep 0 entries the same.

In [4]:
# turn the explicit ratings to implicit feedback data
train_mat = (train_mat > 0).astype(float)
test_mat = (test_mat > 0).astype(float)

## Part 1a: Build the non-personalized recommender (10 points)

In this part, you need to build a non-personalized recommendation model to provide a ranked list of 50 movies as the recommendation for each user. The model is very simple: for each user, the recommendation list is to rank the unwatched movies by their **popularity**, where the popularity is the number of implicit feedback each movie gets. In this case, although it is non-personalized recommender, the recommendation results may be different for users because the unwatched movies are different across users.

In the next cell, write your code to generate the ranked lists of movies by the popularity based recommendation algorithm for every user, store the result in a numpy array of size (#user, 50), entry (u, k) represents the movie id that is ranked at position k in the recommendation list to user u. Print out the top-5 recommended movies and their popularity for the first user (with id 0).

Hint: the popularity can only be calculated from train_mat, you cannot use test_mat here. 

In [13]:
# Generate a ranked list of movies by the popularity based recommendation algorithm. And print out the id and popularity of the top5 movies for the first user.
# Your Code Here...
popularity = np.sum(train_mat, axis=0, dtype=int)

recommendation_non = np.empty((num_user, 50), dtype=int)

for u in range(num_user):
  rec_movie = np.where(train_mat[u, :] == 0)[0]
  rec_score = popularity[rec_movie]
  recommendation_non[u] = rec_movie[np.argsort(rec_score)][-50:][::-1]

print("ID\tPopularity")
for i in recommendation_non[0, :5]:
  print(f"{i}\t{popularity[i]}")

ID	Popularity
104	2423
124	2086
64	2064
113	1862
97	1845


## Part 1b: Evaluate the non-personalized recommender (10 points)

In this part, you need to evaluate your non-personalized recommendation by the held-out testing dataset test_mat for each user. For the implicit recommendation, two typical metrics are recall@k and precision@k. Here, you need to write the code to calculate recall@k and  precision@k for k=5, 20, 50 for each user, i.e., six metrics for every user. And please print out the average over all users for these six metrics.

Hint: if a user does not have any testing samples in test_mat, do not include this user in the final averaged metric.


In [22]:
# Calculate recall@k and precision@k with k=5, 20, 50 and print out the average over all users for these 9 metrics.
# Your Code Here...

def evaluation(test_mat, recommendation):
  recall5 = np.zeros(num_user)
  recall20 = np.zeros(num_user)
  recall50 = np.zeros(num_user)
  precision5 = np.zeros(num_user)
  precision20 = np.zeros(num_user)
  precision50 = np.zeros(num_user)
  user = []

  for u in range(test_mat.shape[0]):
      test_pos = np.where(test_mat[u, :] > 0)[0]
      test_num = len(test_pos)
      if test_num:
        user.append(u)
        rec_u = recommendation[u]
        for i, m in enumerate(rec_u):
          if m in test_pos:
            if i < 5:
              recall5[u] += 1
              precision5[u] += 1
            if i < 20:
              recall20[u] += 1
              precision20[u] += 1
            if i < 50:
              recall50[u] += 1
              precision50[u] += 1
        recall5[u] /= test_num
        recall20[u] /= test_num
        recall50[u] /= test_num
        precision5[u] /= 5
        precision20[u] /= 20
        precision50[u] /= 50
  
  avg_recall5 = recall5[user].sum()/len(user)
  avg_recall20 = recall20[user].sum()/len(user)
  avg_recall50 = recall50[user].sum()/len(user)
  avg_precision5 = precision5[user].sum()/len(user)
  avg_precision20 = precision20[user].sum()/len(user)
  avg_precision50 = precision50[user].sum()/len(user)

  print('recall@5\t[%.6f],\t||\t recall@20\t[%.6f],\t||\t recall@50\t[%.6f]' % (avg_recall5, avg_recall20, avg_recall50))
  print('precision@5\t[%.6f],\t||\t precision@20\t[%.6f],\t||\t precision@50\t[%.6f]' % (avg_precision5, avg_precision20, avg_precision50))
  print('')


In [23]:
evaluation(test_mat, recommendation_non)

recall@5	[0.037861],	||	 recall@20	[0.108553],	||	 recall@50	[0.193615]
precision@5	[0.273013],	||	 precision@20	[0.210389],	||	 precision@50	[0.157815]



# Part 2. Build the Collaborative Filtering Model for Implicit Feedback (25 points)

In this part, we will need to build **personalized** models instead of non-personalized models as in Part 1. We study how collaborative filtering algorithms work for recommendations with implicit feedback. The overall pipeline is the same as in Part 1. But now you need to implement a user-user collaborative filtering algorithm for recommendation. You will also evaluate your personalized models to compare them with the non-personalized one in Part 1.

In this part, you will use the same MovieLens 1M dataset, and:

* write the code to implement a user-user collaborative filtering algorithm,
* and evaluate your recommender.

First, we need to load and preprocess the experiment dataset. We still use the MovieLens 1M data from https://grouplens.org/datasets/movielens/1m/ in this part.

Then, you need to write code to implement a user-user collaborative filtering algorithm with **cosine similarity**. Although we have not discussed in class how to use collaborative filtering for implicit feedback, it is quite straightforward. 

* We first need to calculate the cosine similarity between users based on the binary feedback vectors of users; 
* Then, for a specific user $u$, we predict a preference vector for the user $u$ by weighted averaging the binary feedback vectors of N users who are most similar to $u$;
* And last, rank the movies based on the predicted preference vector of the user $u$ as recommendations. 

The predicted preference score from user $u$ to movie $i$ can be calculated as: $p_{u,i}=\frac{\sum_{u^\prime\in N}s(u,u^\prime)r_{u^\prime,i}}{\sum_{u^\prime\in N}|s(u,u^\prime)|}$, where $s(u,u^\prime)$ is the cosine similarity, and we set the size of $N$ as 10.

In the next cell, write your code to generate the ranked lists of movies by this user-user collaborative filtering recommendation algorithm for every user, store the result in a numpy array of size (#user, 50), where entry (u, k) represents the movie id that is ranked at position k in the recommendation list to user u. 

* Hint: for a user $u$, the movies user $u$ liked in the train_mat should be excluded in the top 50 recommendation list

In [7]:
# Your Code Here...
N = 10
cos_similarity = np.empty((num_user, N))
similar_user = np.empty((num_user, N), dtype=int)

for u in range(num_user):
  s = np.inner(train_mat[u], train_mat) / (np.linalg.norm(train_mat[u]) * np.linalg.norm(train_mat, axis=1))
  s[u] = float("-inf")
  similar_user[u] = np.argsort(s)[-10:]
  cos_similarity[u] = s[similar_user[u]]


recommendation_uucf = np.empty((num_user, 50), dtype=int)

for u in range(num_user):
  rec_movie = np.where(train_mat[u, :] == 0)[0]
  rec_score = np.inner(cos_similarity[u], train_mat[similar_user[u], rec_movie.reshape(-1,1)]) / np.absolute(cos_similarity[u]).sum()
  recommendation_uucf[u] = rec_movie[np.argsort(rec_score)][-50:][::-1]

Last, you need to evaluate your user-user collaborative filtering algorithm by the held-out testing dataset test_mat for each user. Here, we use the same metrics recall@k and precision@k we used in Part 2c of Module 1. So you can use the same code here to calculate recall@k and precision@k for k=5, 20, 50 for each user, i.e., six metrics for every user. And please print out the average over all users for these six metrics.

In [24]:
# Calculate recall@k, precision@k with k=5, 20, 50 and print out the average over all users for these 6 metrics.
# Your Code Here...
evaluation(test_mat, recommendation_uucf)

recall@5	[0.070648],	||	 recall@20	[0.191710],	||	 recall@50	[0.321898]
precision@5	[0.378477],	||	 precision@20	[0.290886],	||	 precision@50	[0.221715]



**Question (5 points):** do you observe a better result compared with models implemented in the previous part? What's reason do you think that brings this improvement?

Write your answer here:

The result shows that the user-user collaborative filtering model perfoms better than the non-personalized recommender. This is because that people have their own taste in movies, but the non-personalized model only recommend movies based on the popularity. On the other hand, the UUCF model finds similar users and recommend movies based on their interest.

|       | recall@5 | precision@5 | recall@20 | precision@20 | recall@50 | precision@50 |
| --- | :-: | :-: | :-: | :-: | :-: | :-: |
| Non-Personalized Recommender | 0.038 | 0.273 | 0.109 | 0.210 | 0.194 | 0.158 |
| Collaborative Filtering Model | **0.071** | **0.378** | **0.192** | **0.291** | **0.322** | **0.222** |

# Part 3. Build the Matrix Factorization Model for Implicit Feedback (25 points)

Now we turn to study how matrix factorization algorithms works for recommendations with implicit feedback. The overall pipeline is the same as in Part 1 and Part 2. But now you need to implement a matrix factorization algorithm for recommendation. You will also compare your matrix factorization model to the non-personalized one in Part 1 and collaborative filtering one in Part 2.

In this part, you will use the same MovieLens 1M dataset, and:
* write the code to implement a matrix factorization algorithm,
* and evaluate your recommender.

First, let's recap the matrix factorization model introduced in class. The MF model can be mathematically represented as: 

<center>$\underset{\mathbf{P},\mathbf{Q}}{\text{min}}\,\,L=\sum_{(u,i)\in\mathcal{O}}(\mathbf{P}_u\cdot\mathbf{Q}^\top_i-r_{u,i})^2+\lambda(\lVert\mathbf{P}\rVert^2_{\text{F}}+\lVert\mathbf{Q}\rVert^2_{\text{F}})$,</center>
    
where $\mathbf{P}$ is the user latent factor matrix of size (#user, #latent); $\mathbf{Q}$ is the movie latent factor matrix of size (#movie, #latent); $\mathcal{O}$ is a user-movie pair set containing all user-movie pairs having ratings in train_mat; $r_{u,i}$ represents the rating for user u and movie i; $\lambda(\lVert\mathbf{P}\rVert^2_{\text{F}}+\lVert\mathbf{Q}\rVert^2_{\text{F}})$ is the regularization term to overcome overfitting problem, $\lambda$ is the regularization weight (a hyper-parameter manually set by developer, i.e., you), and $\lVert\mathbf{P}\rVert^2_{\text{F}}=\sum_{x}\sum_{y}(\mathbf{P}_{x,y})^2$, $\lVert\mathbf{Q}\rVert^2_{\text{F}}=\sum_{x}\sum_{y}(\mathbf{Q}_{x,y})^2$. Such an L function is called the **loss function** for the matrix factorization model. The goal of training an MF model is to find appropriate $\mathbf{P}$ and $\mathbf{Q}$ to minimize the loss L.

Although we have not discussed in class how to use matrix factorization algorithm for implicit feedback, it is quite straightforward. The main idea is the same as the MF model for explicit ratings. But instead of predicting explicit ratings, here the MF is to predict binary ratings based on which movies are ranked for users.

The only challenge now is that in an implicit feedback dataset, there is only positive signal (i.e., '1' in the train_mat) without negative signal. Hence, to let our MF work for this implicit feedback, a simple but powerful method -- random negative sampling -- is adopted. The main idea is that in each training epoch, we randomly sample user-movie pairs without positive feedback in train_mat (user-movie pairs with '0' in train_mat) to be negative feedback, and mix them with positive feedback as the training data to train the MF model. The **negative sampling method is already provided**.

In the Next cell, you need to complete the MF_implicit class. There are five functions in this class: 

* 'init' (**provided**) is to initialize the variables the MF class needs, which takes 5 inputs: train_mat, test_mat, latent, lr, and reg. 'train_mat' and 'test_mat' are the corresponfing training and testing matrices we have. 'latent' represents the latent dimension we set for the MF model. 'lr' represents the learning rate, i.e., the update step in each optimization iteration, default is 0.01. 'reg' represents the regularization weight, i.e., the $\lambda$ in the MF formulation. 

* 'negative_sampling' (**provided**) is to do the random negative sampling to generate negative signals for training the MF. It returns a list of users and a list of movies as the training samples, mixing positive and negative user-movie pairs.

* 'test' (**provided**) is to evaluate the trained MF on test_mat and print out recall@k and precision@k. 

* 'predict' (**need to be completed**) is to generate the ranked lists of movies by the trained MF for every user, store the ranking result (named 'recommendation') in a numpy array of size (#user, 50), where entry (u, k) represents the movie id that is ranked at position k in the recommendation list to user u. Return the 'recommendation' variable. 

* 'train' (**need to be completed**) is to train the MF model. There is only one input to this function: an int variable 'epoch' to indicate how many epochs for training the model. The main logic of this function is the same as the one you already implemented in Part 1a. The main body of this function should be a loop for 'epoch' iterations. In each iteration, following the algorithm to update the MF model:

        1. Call 'negative_sampling()' to generate a list of users and a list of movies mixing positive and negative user-movie pairs as training samples.
        2. Randomly shuffle training user-movie pairs  (i.e., user-movie pairs from step 1)
        2. Have an inner loop to iterate each user-movie pair:
                a. given a user-movie pair (u,i), update the user latent factor and movie latent factor by gradient decsent:    
<center>$\mathbf{P}_u=\mathbf{P}_u-\gamma [2(\mathbf{P}_u\cdot\mathbf{Q}_i^\top-r_{u,i})\cdot\mathbf{Q}_i+2\lambda\mathbf{P}_u]$</center>    
<center>$\mathbf{Q}_i=\mathbf{Q}_i-\gamma [2(\mathbf{P}_u\cdot\mathbf{Q}_i^\top-r_{u,i})\cdot\mathbf{P}_u+2\lambda\mathbf{Q}_i]$</center>    
<center>where $\mathbf{P}_u$ and $\mathbf{Q}_i$ are row vectors of size (1, #latent), $\gamma$ is learning rate (default is 0.01), $\lambda$ is regularization weight, and $r_{u,i}$ now takes binary value.</center>
        
        3. After iterating over all user-movie pairs, call 'test()' to evaluate the current MF model.

**Note: you are not supposed to delete or modify the provided code.**

**Note: for this part, it is necessary to read and understand the provided code, because you will need to call the provided functions in your code.**

In [ ]:
class MF_implicit:
    def __init__(self, train_mat, test_mat, latent=5, lr=0.01, reg=0.01):
        self.train_mat = train_mat  # the training rating matrix of size (#user, #movie)
        self.test_mat = test_mat  # the training rating matrix of size (#user, #movie)
        
        self.latent = latent  # the latent dimension
        self.lr = lr  # learning rate
        self.reg = reg  # regularization weight, i.e., the lambda in the objective function
        
        self.num_user, self.num_movie = train_mat.shape
        
        self.sample_user, self.sample_movie = self.train_mat.nonzero()  # get the user-movie paris having ratings in train_mat
        self.num_sample = len(self.sample_user)  # the number of user-movie pairs having ratings in train_mat

        self.user_test_like = []
        for u in range(self.num_user):
            self.user_test_like.append(np.where(self.test_mat[u, :] > 0)[0])

        self.P = np.random.random((self.num_user, self.latent))  # latent factors for users, size (#user, self.latent), randomly initialized
        self.Q = np.random.random((self.num_movie, self.latent))  # latent factors for users, size (#movie, self.latent), randomly initialized
        
    def negative_sampling(self):
        negative_movie = np.random.choice(np.arange(self.num_movie), size=(len(self.sample_user)), replace=True)
        true_negative = self.train_mat[self.sample_user, negative_movie] == 0
        negative_user = self.sample_user[true_negative]
        negative_movie = negative_movie[true_negative]
        return np.concatenate([self.sample_user, negative_user]), np.concatenate([self.sample_movie, negative_movie])

    def train(self, epoch=20):
        """
        Goal: Write your code to train your matrix factorization model for epoch iterations in this function
        Input: epoch -- the number of training epoch 
        """
        for ep in range(epoch):
            """ 
            Write your code here to implement the training process for one epoch, 
            at the end of each epoch, run self.test() to evaluate current version of MF.
            """
            train_user, train_movie = self.negative_sampling()
            shuffle = np.arange(train_user.shape[0])
            np.random.shuffle(shuffle)

            for idx in shuffle:
              u = train_user[idx]
              i = train_movie[idx]
              pu = self.P[u]
              qi = self.Q[i]
              r = 1 if idx < self.num_sample else 0
              self.P[u] = pu - self.lr * (2 * (np.inner(pu, qi) - r) * qi + 2 * self.reg * pu)
              self.Q[i] = qi - self.lr * (2 * (np.inner(pu, qi) - r) * pu + 2 * self.reg * qi)
            
            print(f"epoch = {ep+1} ======================\n")
            self.test()
            """
            End of your code for this function
            """
            
    def predict(self):
        """
        Write your code here to implement the prediction function, which generates the ranked lists of movies 
        by the trained MF for every user, store the result (named 'recommendation') in a numpy array of size (#user, 50), where entry (u, k) 
        represents the movie id that is ranked at position k in the recommendation list to user u. Return the 'recommendation' variable. 
        """
        recommendation = np.empty((self.num_user, 50), dtype=int)
        for u in range(self.num_user):
          rec_movie = np.where(self.train_mat[u] == 0)[0]
          rec_score = np.inner(self.P[u], self.Q[rec_movie])
          recommendation[u] = rec_movie[np.argsort(rec_score)][-50:][::-1]
        
        """
        End of your code for this function
        """
        return recommendation
    
    def test(self):
        recommendation = self.predict()

        recalls = np.zeros(3)
        precisions = np.zeros(3)
        user_count = 0.

        for u in range(self.num_user):
            test_like = self.user_test_like[u]
            test_like_num = len(test_like)
            if test_like_num == 0:
                continue
            rec = recommendation[u, :]
            hits = np.zeros(3)
            for k in range(50):
                if rec[k] in test_like:
                    if k < 50:
                        hits[2] += 1
                        if k < 20:
                            hits[1] += 1
                            if k < 5:
                                hits[0] += 1
            recalls[0] += (hits[0] / test_like_num)
            recalls[1] += (hits[1] / test_like_num)
            recalls[2] += (hits[2] / test_like_num)
            precisions[0] += (hits[0] / 5.)
            precisions[1] += (hits[1] / 20.)
            precisions[2] += (hits[2] / 50.)
            user_count += 1

        recalls /= user_count
        precisions /= user_count

        print('recall@5\t[%.6f],\t||\t recall@20\t[%.6f],\t||\t recall@50\t[%.6f]' % (recalls[0], recalls[1], recalls[2]))
        print('precision@5\t[%.6f],\t||\t precision@20\t[%.6f],\t||\t precision@50\t[%.6f]' % (precisions[0], precisions[1], precisions[2]))
        print('')

Now, run the next cell to build and train your implemented MF model for implicit feedback. The expected time used for training one epoch is 20s to 2 min.

In [ ]:
mf_implicit = MF_implicit(train_mat, test_mat, latent=5, lr=0.01, reg=0.0001)
mf_implicit.train(epoch=20)

epoch = 1 ======================

recall@5	[0.012405],	||	 recall@20	[0.052740],	||	 recall@50	[0.129528]
precision@5	[0.134503],	||	 precision@20	[0.132765],	||	 precision@50	[0.122854]

epoch = 2 ======================

recall@5	[0.022735],	||	 recall@20	[0.079082],	||	 recall@50	[0.161196]
precision@5	[0.206325],	||	 precision@20	[0.174851],	||	 precision@50	[0.143338]

epoch = 3 ======================

recall@5	[0.030400],	||	 recall@20	[0.090201],	||	 recall@50	[0.173676]
precision@5	[0.239470],	||	 precision@20	[0.187003],	||	 precision@50	[0.148891]

epoch = 4 ======================

recall@5	[0.025901],	||	 recall@20	[0.091360],	||	 recall@50	[0.176319]
precision@5	[0.220364],	||	 precision@20	[0.192988],	||	 precision@50	[0.151063]

epoch = 5 ======================

recall@5	[0.030409],	||	 recall@20	[0.093244],	||	 recall@50	[0.177132]
precision@5	[0.247450],	||	 precision@20	[0.197757],	||	 precision@50	[0.154858]

epoch = 6 ======================

recall@5	[0.027048],	||	 r

**Question (5 points):** do you observe a better result compared with models implemented in previous two parts? What's reason do you think that brings this improvement?

Write your answer here:

The matrix factorization model performs better than non-personalized model. This is because the MF model recommends movies based on the latent factors whihch represent users' preference and items' genre.

However, the MF model performs worse than the user-user collaborative filtering model. The reason might be that the negative sample chooses the movies that the user u is interested in. It would get low score because it is negative samelpe. The other reason might be the training epoches are not enough.

|       | recall@5 | precision@5 | recall@20 | precision@20 | recall@50 | precision@50 |
| --- | :-: | :-: | :-: | :-: | :-: | :-: |
| Non-Personalized Recommender | 0.038 | 0.273 | 0.109 | 0.210 | 0.194 | 0.158 |
| Collaborative Filtering Model | **0.071** | **0.378** | **0.192** | **0.291** | **0.322** | **0.222** |
| Matrix Factorization Model | 0.038 | 0.289 | 0.123 | 0.243 | 0.241 | 0.201 |

# Part 4: Build the Bayesian Personalized Ranking (BPR) Model with Implicit Feedback (20 points)

In this first part, you will need to build a **Bayesian Personalized Ranking (BPR)** model. You will also evaluate your BPR model to compare with the non-personalized one in Part 1, neighborhood-based collaborative filtering model in Part 2, and matrix factorization model in Part 3. 

For this part, we will:

* build a BPR model,
* and evaluate your recommender.

We still use the MovieLens 1M data from https://grouplens.org/datasets/movielens/1m/ in this part, and use the same loading and preprocessing process.

First, let's implement the Bayesian Personalized Ranking model introduced in class. The BPR model can be mathematically represented as: 

<center>$\underset{\mathbf{P},\mathbf{Q}}{\text{max}}\,\,L=\sum_{(u,i,j)\in\mathcal{O}}\sigma(\mathbf{P}_u\cdot\mathbf{Q}^\top_i-\mathbf{P}_u\cdot\mathbf{Q}^\top_j)+\lambda(\lVert\mathbf{P}\rVert^2_{\text{F}}+\lVert\mathbf{Q}\rVert^2_{\text{F}})$,</center>
    
where $\mathbf{P}$ is the user latent factor matrix of size (#user, #latent); $\mathbf{Q}$ is the movie latent factor matrix of size (#movie, #latent); $\sigma(\cdot)$ is the Sigmoid function; $\mathcal{O}$ is a (user, positive movie, negative movie) tuple set, and each tuple $(u,i,j)$ is a training sample with user $u$ and a posotive movie $i$ with value 1 in train_mat and a negative movie $j$ with value 0 in train_mat; $\lambda(\lVert\mathbf{P}\rVert^2_{\text{F}}+\lVert\mathbf{Q}\rVert^2_{\text{F}})$ is the regularization term to overcome overfitting problem, $\lambda$ is the regularization weight (a hyper-parameter manually set by the developer, i.e., you), and $\lVert\mathbf{P}\rVert^2_{\text{F}}=\sum_{x}\sum_{y}(\mathbf{P}_{x,y})^2$, $\lVert\mathbf{Q}\rVert^2_{\text{F}}=\sum_{x}\sum_{y}(\mathbf{Q}_{x,y})^2$. Such an L function is called the **loss function** for the matrix factorization model. The goal of training a BPR model is to find appropriate $\mathbf{P}$ and $\mathbf{Q}$ to **maximize** the loss L.

To implement such a BPR model, here we will write a Python class for the model. Similar to the implicit_MF model you have already implemented in Part 3, there are five functions in this BPR class: 

* The 'init' function (**provided**) is to initialize the variables the BPR class needs, which takes 5 inputs: train_mat, test_mat, latent, lr, and reg. 'train_mat' and 'test_mat' are the corresponfing training and testing matrices we have. 'latent' represents the latent dimension we set for the BPR model. 'lr' represents the learning rate, i.e., the update step in each optimization iteration, default is 0.01. 'reg' represents the regularization weight, i.e., the $\lambda$ in the BPR formulation.

* 'negative_sampling' (**provided**) is to do the random negative sampling to generate negative signals for training the BPR. It returns a list of users, a list of positive movies of these users, and a list of negative movies of these user. These three lists form the training set $\mathcal{O}$.

* 'test' (**provided**) is to evaluate the trained MF on test_mat and print out recall@k and precision@k. 

* 'predict' (**provided**) is to generates the ranked lists of movies by the trained MF for every user, store the ranking result (named 'recommendation') in a numpy array of size (#user, 50), where entry (u, k) represents the movie id that is ranked at position k in the recommendation list to user u. Return the 'recommendation' variable. 

* 'train' (**need to be completed**) is to train the BPR model. There is only one input to this function: an int variable 'epoch' to indicate how many epochs for training the model. The main process of one training epoch is:

        1. Call 'negative_sampling()' to generate training samples: a list of users, a list of positive movies of these users, and a list of negative movies of these users.
        2. Randomly shuffle training samples from step 1.
        3. Have an inner loop to iterate each (user, positive movie, negative movie) tuple in the shuffled training samples:
                a. given the current training tuple -- a user u, a positive movie i, and a negative movie j, update the user latent factor and movie latent factor by gradient decsent:
<center>$\mathbf{P}_u=\mathbf{P}_u - \gamma[-\frac{e^{-\widehat{x}}}{1 + e^{-\widehat{x}}} (\mathbf{Q}_i - \mathbf{Q}_j) + \lambda\mathbf{P}_u]$</center>
<center>$\mathbf{Q}_i=\mathbf{Q}_i - \gamma[-\frac{e^{-\widehat{x}}}{1 + e^{-\widehat{x}}} \mathbf{P}_U + \lambda\mathbf{Q}_i]$</center>
<center>$\mathbf{Q}_j=\mathbf{Q}_j - \gamma[\frac{e^{-\widehat{x}}}{1 + e^{-\widehat{x}}} \mathbf{P}_U + \lambda\mathbf{Q}_j]$</center>   
<center>where $\mathbf{P}_u$, $\mathbf{Q}_i$, and $\mathbf{Q}_j$ are of size (1, #latent), $\gamma$ is learning rate (default is 0.01), $\lambda$ is regularization weight, and $\widehat{x}=\mathbf{P}_u\cdot\mathbf{Q}^\top_i-\mathbf{P}_u\cdot\mathbf{Q}^\top_j$.</center>
        
        3. After iterate all training samples, call 'test()' to evaluate the current BPR model.


In the next cell, you will need to fill in the 'train' function based on the description above. 

*Hint that similar to the training process of MF in Part 3, in each update iteration, you need to use the old P and Q from last update iteration to update P and Q in current update iteration.

**NOTE that you should not delete or modify the provided code.**

In [ ]:
import math

class BPR:
    def __init__(self, train_mat, test_mat, latent=5, lr=0.01, reg=0.01):
        self.train_mat = train_mat  # the training rating matrix of size (#user, #movie)
        self.test_mat = test_mat  # the training rating matrix of size (#user, #movie)
        
        self.latent = latent  # the latent dimension
        self.lr = lr  # learning rate
        self.reg = reg  # regularization weight, i.e., the lambda in the objective function
        
        self.num_user, self.num_movie = train_mat.shape
        
        self.positive_user, self.positive_movie = self.train_mat.nonzero()  # get the user-movie paris having ratings in train_mat

        self.user_test_like = []
        for u in range(self.num_user):
            self.user_test_like.append(np.where(self.test_mat[u, :] > 0)[0])

        self.P = np.random.random((self.num_user, self.latent))  # latent factors for users, size (#user, self.latent), randomly initialized
        self.Q = np.random.random((self.num_movie, self.latent))  # latent factors for users, size (#movie, self.latent), randomly initialized
        
    def negative_sampling(self): 
        # do the negative sampling for each of the positive user-movie pair. Here we set negative sampling rate as 2, 
        # i.e., for each positive user-movie pair, randomly sample two negative movies. 
        # This function returns final training data: a user list, a positive movie list, and a negative movie list
        sample_user = np.tile(self.positive_user, 2)
        sample_pos_movie = np.tile(self.positive_movie, 2)
        sample_neg_movie = np.random.choice(np.arange(self.num_movie), size=(len(sample_user)), replace=True)
        true_negative = self.train_mat[sample_user, sample_neg_movie] == 0
        return sample_user[true_negative], sample_pos_movie[true_negative], sample_neg_movie[true_negative]

    def train(self, epoch=20):
        """
        Goal: Write your code to train your matrix factorization model for epoch iterations in this function
        Input: epoch -- the number of training epoch 
        """
        for ep in range(epoch):
            """ 
            Write your code here to implement the training process for one epoch, 
            at the end of each epoch, run self.test() to evaluate current version of BPR.
            """
            sample_user, sample_pos_movie, sample_neg_movie, = self.negative_sampling()
            shuffle = np.arange(sample_user.shape[0])
            np.random.shuffle(shuffle)

            for idx in shuffle:
              u, i, j = sample_user[idx], sample_pos_movie[idx], sample_neg_movie[idx]
              pu, qi, qj = self.P[u], self.Q[i], self.Q[j]
              x = np.inner(pu, qi - qj)
              dsig = math.exp(-x) / (1 + math.exp(-x))
              self.P[u] = pu - self.lr * (-1 * dsig * (qi - qj) + self.reg * pu)
              self.Q[i] = qi - self.lr * (-1 * dsig * pu + self.reg * qi)
              self.Q[j] = qj - self.lr * (dsig * pu + self.reg * qj)

            print(f"epoch = {ep+1} ======================\n")
            self.test()

            """
            End of your code for this function
            """
            
    def predict(self):
        # The prediction function, which generates the ranked lists of movies 
        # by the trained BPR for every user, store the result (named 'recommendation') in a numpy array of size (#user, 50), where entry (u, k) 
        # represents the movie id that is ranked at position k in the recommendation list to user u. Return the 'recommendation' variable. 
        prediction_mat = np.matmul(self.P, self.Q.T)
        recommendation = []
        for u in range(self.num_user):
            scores = prediction_mat[u]
            train_like = np.where(train_mat[u, :] > 0)[0]
            scores[train_like] = -9999
            top50_iid = np.argpartition(scores, -50)[-50:]
            top50_iid = top50_iid[np.argsort(scores[top50_iid])[-1::-1]]
            recommendation.append(top50_iid)
        recommendation = np.array(recommendation)
        return recommendation
    
    def test(self):
        recommendation = self.predict()

        recalls = np.zeros(3)
        precisions = np.zeros(3)
        user_count = 0.

        for u in range(self.num_user):
            test_like = self.user_test_like[u]
            test_like_num = len(test_like)
            if test_like_num == 0:
                continue
            rec = recommendation[u, :]
            hits = np.zeros(3)
            for k in range(50):
                if rec[k] in test_like:
                    if k < 50:
                        hits[2] += 1
                        if k < 20:
                            hits[1] += 1
                            if k < 5:
                                hits[0] += 1
            recalls[0] += (hits[0] / test_like_num)
            recalls[1] += (hits[1] / test_like_num)
            recalls[2] += (hits[2] / test_like_num)
            precisions[0] += (hits[0] / 5.)
            precisions[1] += (hits[1] / 20.)
            precisions[2] += (hits[2] / 50.)
            user_count += 1

        recalls /= user_count
        precisions /= user_count

        print('recall@5\t[%.6f],\t||\t recall@20\t[%.6f],\t||\t recall@50\t[%.6f]' % (recalls[0], recalls[1], recalls[2]))
        print('precision@5\t[%.6f],\t||\t precision@20\t[%.6f],\t||\t precision@50\t[%.6f]' % (precisions[0], precisions[1], precisions[2]))
        print('')

Now, let's train a BPR model based on your implementation. The code is provided, you just need to execute the next cell. The expectations are: 

* first, the code can be successfully executed without error; 
* and second, the recall@k and precision@k on **test_mat** of each training epoch should be printed out for all 20 epochs.


* Hint: the expected time used for training is around 1 minute to 5 minutes per training epoch.

In [ ]:
bpr = BPR(train_mat, test_mat, latent=5, lr=0.01, reg=0.001)
bpr.train(epoch=20)

epoch = 1 ======================

recall@5	[0.034968],	||	 recall@20	[0.105035],	||	 recall@50	[0.188376]
precision@5	[0.258477],	||	 precision@20	[0.206465],	||	 precision@50	[0.154798]

epoch = 2 ======================

recall@5	[0.034815],	||	 recall@20	[0.105514],	||	 recall@50	[0.189177]
precision@5	[0.260166],	||	 precision@20	[0.206283],	||	 precision@50	[0.155222]

epoch = 3 ======================

recall@5	[0.035541],	||	 recall@20	[0.105341],	||	 recall@50	[0.188762]
precision@5	[0.263278],	||	 precision@20	[0.206821],	||	 precision@50	[0.155361]

epoch = 4 ======================

recall@5	[0.035530],	||	 recall@20	[0.105910],	||	 recall@50	[0.190542]
precision@5	[0.265099],	||	 precision@20	[0.207980],	||	 precision@50	[0.157023]

epoch = 5 ======================

recall@5	[0.036079],	||	 recall@20	[0.106274],	||	 recall@50	[0.192402]
precision@5	[0.267384],	||	 precision@20	[0.210248],	||	 precision@50	[0.159563]

epoch = 6 ======================

recall@5	[0.036504],	||	 r

**Question (5 points):** do you observe a better result compared with models implemented in previous three parts? What's reason do you think that brings this improvement?

Write your answer here:

It is no suprise that the bayesian personalized ranking model performs better than non-personalized recommender. This is because BPR considers the users' interest and items' genre.

The result of BPR and the result of MF are close Also, BPR performs worse than the UUCF model. In my opion, the reason might be the assumption of BPR. The BPR model assumes that observed user-item pairs are preferred over unobserved user-item pairs. However, the user might be interested in that item.


|       | recall@5 | precision@5 | recall@20 | precision@20 | recall@50 | precision@50 |
| --- | :-: | :-: | :-: | :-: | :-: | :-: |
| Non-Personalized Recommender | 0.038 | 0.273 | 0.109 | 0.210 | 0.194 | 0.158 |
| Collaborative Filtering Model | **0.071** | **0.378** | **0.192** | **0.291** | **0.322** | **0.222** |
| Matrix Factorization Model | 0.038 | 0.289 | 0.123 | 0.243 | 0.241 | 0.201 |
| Bayesian Personalized Ranking | 0.048 | 0.348 | 0.138 | 0.267 | 0.250 | 0.206

# Part 5. Cool Extension! (10 points)

Just like in the first homework, now is your chance to try an interesting extension. 

According to the above results, the user-user collaborative filtering model gets the best recall and the best precision. Therefore, I bulit a hybrid model which combines UUCF and MF. I used the P matrix and the Q matrix in MF model and the cosine similarity in UUCF. The recommend score is:

<center>$p_{u,i}=w \cdot P_u \cdot Q^\top_i + (1-w)\frac{\sum_{u^\prime\in N}s(u,u^\prime)P{u^\prime} \cdot Q^\top_i }{\sum_{u^\prime\in N}|s(u,u^\prime)|}$</center>

$w$ is the weight, and $w = 0.6$ in my implementation. 

The result below shows that the model which combines UUCF and MF performs better than the origin MF model.


|       | recall@5 | precision@5 | recall@20 | precision@20 | recall@50 | precision@50 |
| --- | :-: | :-: | :-: | :-: | :-: | :-: |
| Non-Personalized Recommender | 0.038 | 0.273 | 0.109 | 0.210 | 0.194 | 0.158 |
| Collaborative Filtering Model | **0.071** | **0.378** | **0.192** | **0.291** | **0.322** | **0.222** |
| Matrix Factorization Model | 0.038 | 0.289 | 0.123 | 0.243 | 0.241 | 0.201 |
| Bayesian Personalized Ranking | 0.048 | 0.348 | 0.138 | 0.267 | 0.250 | 0.206
| MF + UUCF | 0.046 | 0.322 | 0.140 | 0.262 | 0.263 | 0.211

In [ ]:
P = mf_implicit.P
Q = mf_implicit.Q
recommendation_mf_uucf = np.empty((num_user, 50), dtype=int)
for u in range(num_user):
  rec_movie = np.where(train_mat[u, :] == 0)[0]
  rec_score = np.inner(P[u], Q[rec_movie])
  uu_score = np.inner(P[similar_user[u]], Q[rec_movie])
  uu_score = np.matmul(cos_similarity[u], uu_score) / np.absolute(cos_similarity[u]).sum()
  rec_score = 0.6 * rec_score + 0.4 * uu_score
  recommendation_mf_uucf[u] = rec_movie[np.argsort(rec_score)][-50:][::-1]
evaluation(test_mat, recommendation_mf_uucf)

recall@5	[0.046295],	||	 recall@20	[0.139535],	||	 recall@50	[0.262500]
precision@5	[0.321523],	||	 precision@20	[0.262401],	||	 precision@50	[0.210791]



# Collaboration Declarations

** You should fill out your collaboration declarations here.**

**Reminder:** You are expected to complete each homework independently. Your solution should be written by you without the direct aid or help of anyone else. However, we believe that collaboration and team work are important for facilitating learning, so we encourage you to discuss problems and general problem approaches (but not actual solutions) with your classmates. You may post on Piazza, search StackOverflow, etc. But if you do get help in this way, you must inform us by filling out the Collaboration Declarations at the bottom of this notebook.

Example: I found helpful code on stackoverflow at https://stackoverflow.com/questions/11764539/writing-fizzbuzz that helped me solve Problem 2.